enhancements/multi-arch/heterogeneous-architecture-clusters.md

```diff
@@ -0,0 +1,433 @@
+for the heterogeneous releases and also the work needed to support upgrades from a homogeneous release -> heterogeneous release in
+the future.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+- The ClusterVersion API's spec field will have an additional architecture field (to be appropriately named) which would serve as an indicator for
+whether the cluster is homogeneous or heterogeneous. The ClusterVersionStatus API will also have a similar field.
+- The NodePoolSpec API in Hypershift would need to have an architecture field so it can pick the images corresponding to the relevant architecture
+- The imagestreams API would also undergo changes as described in the imagestreams section.
+
+#### Failure Modes
+
+No new failure modes are introduced as part of this feature.
+
+#### Support Procedures
+
+TBD
+
+## Implementation History
+
+A heterogeneous payload was built as a proof of concept and spun up worker nodes using ARM with an x86 control plane on AWS. It worked,
+but here are the OCP features we know won’t work properly in such an environment today which are explained in detail above:
+- Imagestreams - does not fully support manifestlists, they will import a single architecture (by default based on the control plane’s architecture,
+though it can be chosen explicitly) from the manifestlist, so anything referencing an imagestream tag will get the image associated with that
+architecture, regardless of what architecture the node is running.
+- Internal registry - does not support manifestlists (storing or pulling). This also means you cannot use the internal registry for image
+pull-through (proxying) for manifestlist-based content.
+- Builds - do not produce manifestlists. Builds do support use of nodeselectors to steer the build to a node of the appropriate architecture.
+- Console - does some content filtering (helm charts, operators) based on the perceived architecture of the cluster, which will be the control plane
+architecture, so not see all content will be seen as being available.
+- Individual OLM operators - it will be a hit/miss which operators support which architectures or if they support multiple architectures. We don’t
+have specific data on this yet.
+- Upgrades - we don’t have an upgrade graph for these payloads, and we have not yet tested even a manual upgrade from one heterogeneous payload
+to another, though we expect it to work without modification.
+- Oc adm release commands are not capable of building manifestlist based payloads today, so you can’t construct your own payload with the tooling.
+
+## Alternatives
+
+The alternative to a heterogeneous architecture cluster would be having multiple clusters with workloads communicating between each other but this
+would have latency and cost implications not to mention the additional failure mode of cluster connectivity.
+
+## Infrastructure Needed
```

Comment:
Assisted Installer team will be affected when the multiarch payload is introduced. We have started some initial investigations regarding that in https://issues.redhat.com/browse/MGMT-8890. Cannot say at the moment if multiarch payload is breaking the Assisted Installer workflow completely, but this is something that should be considered too.

I'm not pushing for writing too much of specific details about AI in this enhancement here, but I'd appreciate at least an acknowledgement that implementation of multiarch payload has a direct impact on the service

Reply:
Last I knew we were still going to have single arch payloads, so other than maybe Multi-arch payloads not being compatible with AI, I wouldn't expect an impact?

Reply:
@mkowalski i'm happy to chat with you offline about the details of https://issues.redhat.com/browse/MGMT-8890, but right now, the heterogeneous payloads are not completely ready to use yet. In this document we have identified phases of support and i can add AI to phase 2/3 depending on the conversation we have, but you mention the oc extract command not working, which we know about and have a ticket to address. Once we have all the bits in place in terms of support and usability, i expect that you should not need anything additional.